include/ocpp/v201/database_handler.hpp

```diff
@@ -178,16 +178,40 @@ class DatabaseHandler : public common::DatabaseHandlerCommon {
     /// charging profiles
 
     /// \brief Inserts or updates the given \p profile to CHARGING_PROFILES table
-    void insert_or_update_charging_profile(const int evse_id, const v201::ChargingProfile& profile);
+    void insert_or_update_charging_profile(
+        const int evse_id, const v201::ChargingProfile& profile,
+        const ChargingLimitSourceEnum charging_limit_source = ChargingLimitSourceEnum::CSO);
 
     /// \brief Deletes the profile with the given \p profile_id
-    void delete_charging_profile(const int profile_id);
+    bool delete_charging_profile(const int profile_id);
+
+    /// \brief Deletes the profiles with the given \p transaction_id
+    void delete_charging_profile_by_transaction_id(const std::string transaction_id);
 
     /// \brief Deletes all profiles from table CHARGING_PROFILES
-    void clear_charging_profiles();
+    bool clear_charging_profiles();
+
+    /// \brief Deletes all profiles from table CHARGING_PROFILES matching \p profile_id or \p criteria
+    bool clear_charging_profiles_matching_criteria(const std::optional<int32_t> profile_id,
+                                                   const std::optional<ClearChargingProfile>& criteria);
+
+    /// \brief Get all profiles from table CHARGING_PROFILES matching \p profile_id or \p criteria
+    std::vector<std::tuple<int, v201::ChargingProfile, ChargingLimitSourceEnum>>
```

Comment:
It might be better to make a helper struct for this return value so that you can actually name the int for example.

Reply:
That would improve it a bit, but I still think a struct is easier to read and maintain. Is there a reason you would like to stick to the tuple?

Reply:
Mainly: It seems odd to add a struct for one specific use case.

It seems I still left the ReportedChargingProfiles struct in tact, so I'll re-use that.

Reply:
Err, that might need to be moved somewhere else because of a recursive import. Where would be a good place to put that struct?

Reply:
We are probably having some other issues when we are running into recursive imports. But that's way out of scope here. Maybe `include/ocpp/v201/types.hpp` is a good spot for it?

Reply:
Moved ReportedChargingProfiles and updated the rest of the code to use it :)